### PR TITLE
[SPARK-24172][SQL] we should not apply operator pushdown to data source v2 many times

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

In `PushDownOperatorsToDataSource`, we use `transformUp` to match `PhysicalOperation` and apply pushdown. This is problematic if we have multiple `Filter` and `Project` above the data source v2 relation.

e.g. for a query
```
Project
  Filter
    DataSourceV2Relation
```

The pattern match will be triggered twice and we will do operator pushdown twice. This is unnecessary, we can use `mapChildren` to only apply pushdown once.

## How was this patch tested?

existing test

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/PushDownOperatorsToDataSource.scala [hunk 1]
```diff
@@ -23,49 +23,46 @@ import org.apache.spark.sql.catalyst.plans.logical.{Filter, LogicalPlan, Project
...
-        case Some(existing) =>
-          existing ++ newFilters
-        case _ =>
-          newFilters
-      }
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    var pushed = false
+    plan transformDown {
+      // PhysicalOperation guarantees that filters are deterministic; no need to check
+      case PhysicalOperation(project, filters, relation: DataSourceV2Relation) if !pushed =>
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/PushDownOperatorsToDataSource.scala]: Is that possible one plan has multiple `PhysicalOperation`?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/PushDownOperatorsToDataSource.scala]: `PhysicalOperation` just accumulates project and filter above a specific node, if we transform down a tree, and only transform once, we will never hit `PhysicalOperation` move than once.

**Comment by rdblue** [comment]
@cloud-fan, I opened #21262 that is similar to this, but does pushdown when converting to a physical plan. You might like that as an alternative because it cleans up `DataSourceV2Relation` quite a bit and adds `output` to the case class arguments like other relations.

The drawback to that approach that I had forgotten about is that it breaks `computeStats` because that runs on the optimized plan (but this affects all the other code paths as well).

Up to you how to continue with this work, I just think we should consider the other approach since it solves a few problems. And `computeStats` is something we should update to work on physical plans anyway, right? Just let me know how you want to move forward. If you want to pull that commit into this PR, I'll close the other one.

**Comment by rdblue** [comment]
+1 (assuming tests pass)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/PushDownOperatorsToDataSource.scala [hunk 2]
```diff
@@ -23,17 +23,10 @@ import org.apache.spark.sql.catalyst.plans.logical.{Filter, LogicalPlan, Project
 import org.apache.spark.sql.catalyst.rules.Rule
 
 object PushDownOperatorsToDataSource extends Rule[LogicalPlan] {
-  override def apply(
-      plan: LogicalPlan): LogicalPlan = plan transformUp {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.mapChildren {
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/PushDownOperatorsToDataSource.scala]: Could you update the PR description (`transformDown` -> `mapChildren`), too?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by kiszk** [comment]
retest this please

**Comment by gatorsmile** [comment]
LGTM Thanks! Merged to master

**Review by gengliangwang — APPROVED** [review]
LGTM

**Pull request closed without merging**